contrib/fedora/cjdns.setgid.patch

```diff
@@ -0,0 +1,143 @@
+             rpcCall0(String_CONST("Security_getUser"), d, ctx, tempAlloc, &ret, true);
+             uid = *Dict_getInt(ret, String_CONST("uid"));
++            group = Dict_getInt(ret, String_CONST("gid"));
+         }
+     } while (0);
+ 
+@@ -477,6 +479,7 @@ static void security(struct Allocator* t
+             Dict* ret = NULL;
+             rpcCall0(String_CONST("Security_getUser"), d, ctx, tempAlloc, &ret, true);
+             uid = *Dict_getInt(ret, String_CONST("uid"));
++            group = Dict_getInt(ret, String_CONST("gid"));
+             int64_t* nka = Dict_getInt(elem, String_CONST("keepNetAdmin"));
+             int64_t* exemptAngel = Dict_getInt(elem, String_CONST("exemptAngel"));
+             keepNetAdmin = ((nka) ? *nka : ((exemptAngel) ? *exemptAngel : 0));
+@@ -525,6 +528,8 @@ static void security(struct Allocator* t
+         Dict_putString(d, String_CONST("root"), String_CONST("/var/run/"), tempAlloc);
+         rpcCall0(String_CONST("Security_chroot"), d, ctx, tempAlloc, NULL, false);
+     }
++    /* FIXME(sdg): moving noforks after setuser might make nproc <- 0,0 work
++     on older kernels, where doing it before causes setuid to fail w EAGAIN. */
+     if (noforks) {
+         Log_debug(log, "Security_noforks()");
+         Dict* d = Dict_new(tempAlloc);
+@@ -534,6 +539,9 @@ static void security(struct Allocator* t
+         Log_debug(log, "Security_setUser(uid:%d, keepNetAdmin:%d)", uid, keepNetAdmin);
+         Dict* d = Dict_new(tempAlloc);
+         Dict_putInt(d, String_CONST("uid"), uid, tempAlloc);
++        if (group) {
++            Dict_putInt(d, String_CONST("gid"), (int)*group, tempAlloc);
++        }
+         Dict_putInt(d, String_CONST("keepNetAdmin"), keepNetAdmin, tempAlloc);
+         rpcCall0(String_CONST("Security_setUser"), d, ctx, tempAlloc, NULL, false);
+     }
+diff -up ./util/Security_admin.c.setgid ./util/Security_admin.c
+--- ./util/Security_admin.c.setgid	2015-08-28 17:32:11.000000000 -0400
++++ ./util/Security_admin.c	2015-10-04 19:30:33.597693363 -0400
+@@ -40,8 +40,10 @@ static void setUser(Dict* args, void* vc
+     struct Jmp jmp;
+     Jmp_try(jmp) {
+         int64_t* user = Dict_getInt(args, String_CONST("uid"));
++        int64_t* group = Dict_getInt(args, String_CONST("gid"));
++        int gid = group ? (int)*group : 0;
+         int64_t* keepNetAdmin = Dict_getInt(args, String_CONST("keepNetAdmin"));
+-        Security_setUser(*user, *keepNetAdmin, ctx->logger, &jmp.handler, requestAlloc);
++        Security_setUser(*user, gid, *keepNetAdmin, ctx->logger, &jmp.handler, requestAlloc);
+     } Jmp_catch {
+         sendError(jmp.message, txid, ctx->admin);
+         return;
+@@ -161,6 +163,7 @@ void Security_admin_register(struct Allo
+     }), admin);
+     Admin_registerFunction("Security_setUser", setUser, ctx, true, ((struct Admin_FunctionArg[]) {
+         { .name = "uid", .required = 1, .type = "Int" },
++        { .name = "gid", .required = 0, .type = "Int" },
```

Comment:
The problem here is that cjd is my name and when you tell someone that something is cjdns, I need to believe that it is secure. So I don't like reviewing tons of patches for different systems, setups, architectures because it all makes me read more code and worry more.